Build Class for Credential Exposure

DJensen94 opened the pull request:
## 🗣 Description ##

Builds out a Class for credential exposure data, making it simpler to access variables from the pages script without passing so many variables

## 💭 Motivation and context ##
This makes the report generator more modular, allowing for data to be accessed anywhere more simply without passing a ton of variables to every script

## 🧪 Testing ##

This current PR needs to pass current pytest /tests

## 📷 Screenshots (if appropriate) ##

<!-- Remove this section and header if not needed -->

## ✅ Checklist ##

- [ ] This PR has an informative and human-readable title.
- [ ] Changes are limited to a single goal - _eschew scope creep!_
- [ ] _All_ future TODOs are captured in issues, which are referenced
      in code comments.
- [ ] All relevant type-of-change labels have been added.
- [ ] I have read the [CONTRIBUTING](../blob/develop/CONTRIBUTING.md) document.
- [ ] These code changes follow [cisagov code standards](https://github.com/cisagov/development-guide).
- [ ] All relevant repo and/or project documentation has been updated
      to reflect the changes in this PR.
- [ ] Tests have been added and/or modified to cover the changes in this PR.
- [ ] All new and existing tests pass.

DJensen94 committed:
Filled in class with same code

using the report metrics code i built out the credentials class will likely need some clean up

DJensen94 committed:
Create name for breach when none is given

cyber six does not always give a breach_name so we created one for it

DJensen94 committed:
rename some variables

Cleaned up some variable names

DJensen94 committed:
Cleaned up appendix function

removed unnecessary code from appendix function

DJensen94 committed:
Add classees for all data sources

Add classes for all the other data sources and metrics

cduhn17 commented:
@DJensen94 
When you get a chance take a look at the LGTM notice. Also be sure to maintain an imperative mood on [commit](https://chris.beams.io/posts/git-commit/#imperative) messages.

DJensen94 committed:
Remove unnecessary values

Remove variables that were not used in creating the return value and Fix incorrect return variable

DJensen94 committed:
Standardize naming convention

Change all variable names to be the same format

DJensen94 committed:
Remove unnecessary variable creation

remove variables that were created just to be passed into a return statement

DJensen94 committed:
Updated Doc strings

Clean up doc strings

DJensen94 committed:
Alphabetize order of functions in each class

Arrange order of class methods alphabetically

DJensen94 committed:
Merge branch 'develop' into Classify_Creds

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
         self.org_uid = org_uid
-        self.query_cyberSix_creds = query_cyberSix_creds(org_uid, start_date, end_date)
+        c6_df = query_cyberSix_creds(org_uid, start_date, end_date)
+        c6_df.loc[
+            c6_df["breach_name"] == "", "breach_name"
+        ] = "Cyber_six_" + pd.to_datetime(c6_df["breach_date"]).dt.strftime("%m/%d/%Y")
```
This line is pretty gnarly and might benefit from being broken up into more than one line of code.

DJensen94 replied:
@dav3r comment has been addressed in the following commit: [d8560b5](https://github.com/cisagov/pe-reports/pull/99/commits/d8560b50ae42d0e6d21fea02bb78526ad7076b4c)

dav3r replied:
That definitely looks better!

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+            c6_df["description"].str.split("Query to find the related").str[0]
+        )
+        c6_df["password_included"] = np.where(c6_df["password"] != "", True, False)
+        self.query_cyberSix_creds = c6_df
```
Does it make any sense to move the cyberSix code above into `query_cyberSix_creds()` or is it specific to this metrics code?

DJensen94 replied:
@dav3r comment has been addressed in the following commit: [d8560b5](https://github.com/cisagov/pe-reports/pull/99/commits/d8560b50ae42d0e6d21fea02bb78526ad7076b4c)
Moved code to db_query.py as suggested

dav3r replied:
@DJensen94 - Just a reminder that our convention is to allow the person who starts a conversation be the one to resolve it (the exception is when you accept commit suggestions and GitHub automatically resolves the conversation).

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
-    def breached(self):
-        """Return total number of breached credentials."""
+class Domains_Masqs:
+    """Domains Masquerading Class."""
```
Unnecessary capitalization:
```suggestion
    """Domains Masquerading class."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+
+    def summary(self):
+        """Return domain masquerading summary information."""
+        df_mal = self.df_mal
```
This line seems unnecessary to me- you can just use `self.df_mal` throughout this function.

DJensen94 replied:
@dav3r, I've addressed this comment in the following commit: [cc99f8b](https://github.com/cisagov/pe-reports/pull/99/commits/cc99f8b7733cb76a25af69a6b49eee2b7cfd9bc9)

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+        assets_df = query_shodan(org_uid, start_date, end_date, "shodan_assets")
+        self.assets_df = assets_df
+
+    def protocol_count(self):
+        """Return a count for each insecure protocol."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+
+        # Horizontal bar: insecure protocol count
+        pro_count = risky_assets.groupby(["protocol"], as_index=False)["protocol"].agg(
+            {"id_count": "count"}
+        )
+        return pro_count
+
+    def risky_assets(self):
+        """Return dataframe of ip and risky protocols."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+        return risky_assets
+
+    def risky_ports_count(self):
+        """Return total count of insecure protocols."""
+        insecure_df = self.insecure_df
+        insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```
You've got a lot of repeated code in these three functions, namely:
```
insecure_df = self.insecure_df
insecure = insecure_df[insecure_df["type"] == "Insecure Protocol"]
insecure = insecure[
    (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
]
risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```
You should consider moving that code to a separate function.

DJensen94 replied:
@dav3r Addressed this comment in the following commit: [051e84b]( https://github.com/cisagov/pe-reports/pull/99/commits/051e84b856e0c6bf33b31d24fcdaf704ac199406)

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+        return verifVulns
+
+    def unverified_cve(self):
+        """Return top 15 unverified cves and their counts."""
```
Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return top 15 unverified CVEs and their counts."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+        return unverif_df
+
+    def unverified_vuln_count(self):
+        """Return the count of ips with unverified vulnerabilities."""
```
Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return the count of IP addresses with unverified vulnerabilities."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+        return len(unverif_df.index)
+
+    def verif_vulns(self):
+        """Return a dataframe with each CVE, the associated ips and the affected ports."""
```
Use proper capitalization for acronyms in docstrings:
```suggestion
        """Return a dataframe with each CVE, the associated IPs and the affected ports."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+
+
+class Cyber_Six:
+    """Dark Web and Cyber Six Data Class."""
```
Remove unnecessary capitalization:
```suggestion
    """Dark web and Cyber Six data class."""
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -12,25 +22,571 @@ def __init__(self, start_date, end_date, org_uid):
+        return dark_web_tags
+
+    def top_cve_table(self):
+        """Get top cves."""
```
Use proper capitalization for acronyms in docstrings:
```suggestion
        """Get top CVEs."""
```

dav3r reviewed (changes requested):
Please take a look at my questions, suggestions, and various minor nitpicks - thanks!

DJensen94 committed:
Update src/pe_reports/metrics.py

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix docstring capitalization

add proper capitalization of acronyms in docstring

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove unnecessary capitalization on doc string

update cyber six class docstring to have proper capitalization

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Fix capitalization for acronyms in docstring

capitalize acronyms in docstring

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Capitalize acronyms in docstring

Capitalized "IPs" in the unverified vuln count method docstring

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Capitalize acronyms in docstring

Capitalized acronyms in unverified_cve method docstring

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Simplify line that fills empty breach_names

Break up code that creates names for breaches when not present and move initial data adjustments to the query function

DJensen94 committed:
Create Staticmethod to remove code redundency

Create a isolate_risky_assets static method to remove duplicate code

DJensen94 committed:
Remove unnecessary variable creation

Remove unnecessary variable creation, use class parameter instead

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -15,22 +24,552 @@ def __init__(self, start_date, end_date, org_uid):
+        insecure = insecure[
+            (insecure["protocol"] != "http") & (insecure["protocol"] != "smtp")
+        ]
+        risky_assets = insecure[["ip", "protocol"]].drop_duplicates(keep="first")
+        return risky_assets
```
Remove unnecessary variable:
```suggestion
        return insecure[["ip", "protocol"]].drop_duplicates(keep="first")
```

dav3r commented on src/pe_reports/metrics.py:
```diff
@@ -15,22 +24,552 @@ def __init__(self, start_date, end_date, org_uid):
+        )
+        return pro_count
+
+    def risky_assets(self):
+        """Return dataframe of ip and risky protocols."""
+        risky_assets = self.isolate_risky_assets(self.insecure_df)
+        return risky_assets
```
Is this function really necessary if all it does is call `self.isolate_risky_assets()` and return its output?

DJensen94 replied:
@dav3r, we can probably remove this and just call the static method (isolate_risky_assets) instead, as long as we are ok to call the static method while using the class

dav3r replied:
I'm fine with that approach.

DJensen94 replied:
@dav3r addressed suggestion in the following commit: [6858068](https://github.com/cisagov/pe-reports/pull/99/commits/68580680776f164eb2e8b144afc603029093c0a5)

dav3r replied:
Thanks!

dav3r reviewed (changes requested):
Thanks for addressing my previous round of feedback- I just have a couple more items based on your recent changes.

DJensen94 committed:
Remove unnecessary variable

directly return dataframe after dropping duplicates instead of creating a new variable

Co-authored-by: dav3r <david.redmin@trio.dhs.gov>

DJensen94 committed:
Remove unnecessary risky_assets function

Remove risky_assets function because  isolate_risky_assets does the same thing

dav3r reviewed (approved):
Looking good here! 👍 👍

cduhn17 reviewed (approved):
Looks good.
